packages/frontend/core/src/components/affine/auth/index.tsx

```diff
@@ -44,58 +49,122 @@ const config: {
   sendEmail: SendEmail,
 };
 
-export const AuthModal: FC<AuthModalBaseProps & AuthProps> = ({
-  open,
-  state,
-  setOpen,
-  email,
-  setAuthEmail,
-  setAuthState,
-  setEmailType,
-  emailType,
-}) => {
-  const onSignedIn = useCallback(() => {
-    setAuthState('signIn');
-    setAuthEmail('');
-    setOpen(false);
-  }, [setAuthState, setAuthEmail, setOpen]);
+export function AuthModal() {
+  const t = useI18n();
+  const [authAtomValue, setAuthAtom] = useAtom(authAtom);
+  const authService = useService(AuthService);
+  const setOpen = useCallback(
+    (open: boolean) => {
+      setAuthAtom(prev => ({ ...prev, openModal: open }));
+    },
+    [setAuthAtom]
+  );
+
+  const signIn = useAsyncCallback(
+    async ({
+      method,
+      payload,
+    }: {
+      method: 'magic-link' | 'oauth';
+      payload: any;
+    }) => {
+      try {
+        switch (method) {
+          case 'magic-link': {
+            const { email, token } = payload;
+            await authService.signInMagicLink(email, token);
+            break;
+          }
+          case 'oauth': {
+            const { code, state } = payload;
+            await authService.signInOauth(code, state);
+            break;
+          }
+        }
+        authService.session.revalidate();
+      } catch (e) {
+        notify.error({
+          title: t['com.affine.client.signin.failed.title'](),
+          message: (e as any).message,
+        });
+      }
+    },
+    [authService, t]
+  );
+
+  useEffect(() => {
+    return events?.ui.onAuthenticationRequest(signIn);
```

Comment:
![6290ecd7e629833e10651a884213c0d.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/T2klNLEk0wxLh4NRDzhk/2e00ea7a-7dbf-4f58-9300-1a6313fda22c.png)

Only the active tab or the tab requesting for login should respond this magic link. Otherwise there might be an issue that user logs in successfully but reports invalid token issue.

Reply:
I think alternatively we can exchange the magic link in main process instead and apply the cookies in electron.

Reply:
I prefer the tab way. Requests in main process won't be shown in devtools and its hard to debug

Reply:
we can use debugging tools for the main process as well btw

Reply:
devtools for renderer is straightforward, and let renderer control the request we can easily achieve state management in the same way as web does